docs/Brew-Livecheck.md

```diff
@@ -0,0 +1,115 @@
+
+* **It's generally preferable to check for versions in the same location as the stable archive, when possible**. This preference is stronger for first-party sources (websites, repositories, etc.) and becomes weaker for third-party sources (e.g. mirrors, another software package manager, etc.).
+
+### Regex guidelines
+
+The regex in a `livecheck` block is used within a strategy to restrict matching to only relevant text (or strings) and to establish what part of the matched text is the version (using a capture group).
+
+Creating a good regex is a balance between being too strict (breaking easily) and too loose (matching more than it should).
+
+* For technical reasons, **the `regex` call in the `livecheck` block should always use parentheses** (e.g. `regex(/example/)`).
+
+* **Regex literals should follow the established Homebrew style**, where the `/.../` syntax is the default and the `%r{...}` syntax is used when a forward slash (`/`) is present.
+
+* **Regexes should adequately represent their intention.** This requires an understanding of basic regex syntax, so we avoid issues like using `.` (match any character) instead of `\.` when we only want to match a period. We also try to be careful about our use of generic catch-alls like  `.*` or `.+`, as it's often better to use something non-greedy and contextually appropriate. For example, if we wanted to match a variety of characters while trying to stay within the bounds of an HTML attribute, we could use something like `[^"' >]*?`.
+
+* **Try not to be too specific in some parts of the regex.** For example, if a file name uses a hyphen (`-`) between the software name and version (e.g. `example-1.2.3.tar.gz`), we may want to use something like `example[._-]v?(\d+(?:\.\d+)+)\.t` instead. This would allow the regex to continue matching if the upstream file name format changes to `example.1.2.3.tar.gz` or `example_1.2.3.tar.gz`.
+
+* **Regexes should be case insensitive unless case sensitivity is explicitly required for matching to work properly**. Case insensitivity is enabled by adding the `i` flag at the end of the regex literal (e.g. `/.../i` or `%r{...}i`). This helps to improve reliability and reduce maintenance, as a case-insensitive regex doesn't need to be manually updated if there are any upstream changes in letter case.
+
+* **Regexes should only use a capturing group around the part of the matched text that corresponds to the version**. For example, in `/href=.*?example-v?(\d+(?:\.\d+)+)(?:-src)?\.t/i`, we're only using a capturing group around the version part (matching a version like `1.2`, `1.2.3`, etc.) and we're using non-capturing groups elsewhere (e.g. `(?:-src)?`). This allows livecheck to rely on the first capture group being the version string.
+
+* **Regexes should only match stable versions**. Regexes should be written to avoid prerelease versions like `1.2-alpha1`, `1.2-beta1`, `1.2-rc1`, etc.
```

Comment:
Is there a regex sample for this?

Reply:
I think most of our existing regexes, except for those Formulae where homebrew-core allows pre-releases (such as`recode` or `vbindiff`) satisfy this criterion.

For example, the regex `/^v?(\d+(?:\.\d+)+)$/i`, which we use when matching git tags avoids matching these pre-release versions. If this example is good enough, it could be added to the doc.

Reply:
This seems like a poor fit for living in every regex and instead something that `livecheck` itself should know to strip out.

Reply:
For what it's worth, I removed this guideline from the most recent draft. How to avoid matching unstable versions depends on the context of a given formula/cask, so it can be hard to explain and may not be something that we can enforce through automation anyway.

---

> This seems like a poor fit for living in every regex and instead something that livecheck itself should know to strip out.

We have some simple logic in `Livecheck#latest_version` that's intended to filter out unstable versions ([logic here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/livecheck/livecheck.rb#L573-L580) and the [`UNSTABLE_VERSION_KEYWORDS` array here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/livecheck/livecheck.rb#L38-L47)). However, this currently only applies when the formula/cask doesn't have a `livecheck` block. Some formulae/casks use an unstable version, so we have to be able to match unstable versions using a `livecheck` block.

In general, `Version` has some built-in handling of alpha/beta/pre/rc, so it seems like this should probably be something that `Version` detects and `livecheck` simply checks. We're not allowed to access the underlying `tokens` of a `Version` (see #7340), so this would require something like `Version#stable?` (and/or `Version#unstable?`), which would check if a version's tokens are one of these unstable types.

In general, I think we can improve the unstable version filtering logic but I don't think we can reasonably avoid the situation where regexes only match stable versions. Though I put forth the idea of only matching stable versions as a guideline, it's honestly just a side effect of the way we have to create regexes for other reasons.

---

When it comes to Git tags, we create a regex that only matches the desired format because we frequently have to filter out all sorts of unexpected/unrelated tags that are perpetually treated as newest (from the standpoint of `Version` comparison). These tags prevent the check from working properly and it's necessary to exclude them. To illustrate why we can't use a looser regex, I'll provide some examples.

Say we have a repository with a stable tag format like `1.2.3`, unstable tags like `1.2.3-rc4`, and a weird tag like `1.2.3-some-fix`. If we use a regex that allows it to match unstable versions (i.e., not restricting the end), it will also match `1.2.3-some-fix` and this will be treated as newest until a later version shows up (e.g., `1.2.4`).

As another example, a number of repositories contain tags for several different pieces of software. Some will use the `1.2.3` tags for the main software in the repository and use a prefix for other versions, like `something-5.6.7`.

If we're loose about the start of the regex, then it would match tags like `1.2.3` as well as `something-5.6.7`. Since we only capture the version part, it will return a version that looks plausible (`5.6.7`) but is actually related to a different piece of software that's not used in the formula/cask. This particular type of problem is only addressed when someone says, "Hm, that doesn't look right", checks upstream, and corrects the issue using a `livecheck` block.

Basically, there are benefits to restricting the format of Git tags and this naturally filters out unstable versions in the process.

---

It's kind of a similar situation with file name regexes. If we have a regex like `/href=.*?example[._-]v?(\d+(?:\.\d+)+)\.t/i`, the trailing `\.t` is primarily intended to only match versions from tarballs. However, a side effect is that it implicitly excludes unstable versions, as it would match `example-1.2.3.tar.gz` but not `example-1.2.4-rc1.tar.gz`.

To match unstable versions, we would need to add something like `[^"' >]*?` at the end of the capturing group, like `(\d+(?:\.\d+)+[^"' >]*?)`. This would match `example-1.2.4-rc1.tar.gz` and produce a version like `1.2.4-rc1`.

It would need to be done this way because the unstable indicator (alpha/beta/rc/pre) must be part of the version text for livecheck to [theoretically] filter it out without using the regex. If we allowed a looser match outside of the capture group (i.e., `/href=.*?example[._-]v?(\d+(?:\.\d+)+)[^"' >]*?\.t/i`), this would also match unstable versions but the returned version would appear to be stable (i.e., `1.2.3-rc4` would appear as `1.2.3`).

As before, this would match more than just unstable versions. If we encounter a weird version that we don't want but it isn't filtered out as an unstable version, then it could erroneously appear as newest to livecheck and essentially break the check.

Since we would have to actively loosen file name regexes to match unstable versions and this would open us up to potential issues, it isn't something that I've seen value in pursuing. To me, it would be more work for a worse result without a compelling reason behind it (i.e., we're generally only interested in stable versions anyway).

There are a small number of formulae/casks using an unstable version and we do actively expand those regexes to be able to match unstable versions. However, if/when the formula/cask returns to a stable version, we would return the regex to a normal format.

---

Does this help to explain things?